subsys/bluetooth/controller/hci_driver.c

```diff
@@ -334,51 +340,59 @@ static void event_packet_process(uint8_t *hci_buf)
 		BT_DBG("Event (0x%02x) len %u", hdr->evt, hdr->len);
 	}
 
-	evt_buf = bt_buf_get_evt(hdr->evt, discardable,
-				 discardable ? K_NO_WAIT : K_FOREVER);
+	evt_buf = bt_buf_get_evt(hdr->evt, discardable, K_NO_WAIT);
 
 	if (!evt_buf) {
 		if (discardable) {
 			BT_DBG("Discarding event");
-			return;
+			return false;
 		}
 
-		BT_ERR("No event buffer available");
-		return;
+		return true;
 	}
 
 	net_buf_add_mem(evt_buf, &hci_buf[0], hdr->len + sizeof(*hdr));
 	bt_recv(evt_buf);
+	return false;
 }
 
 static bool fetch_and_process_hci_msg(uint8_t *p_hci_buffer)
 {
 	int errcode;
-	sdc_hci_msg_type_t msg_type;
-
-	errcode = MULTITHREADING_LOCK_ACQUIRE();
-	if (!errcode) {
-		errcode = hci_internal_msg_get(p_hci_buffer, &msg_type);
-		MULTITHREADING_LOCK_RELEASE();
-	}
-
-	if (errcode) {
-		return false;
+	static sdc_hci_msg_type_t msg_type;
+	/*msg_type == 0 means we should fetch a new event from sdc.*/
+	if (msg_type == 0)
+	{
+		errcode = MULTITHREADING_LOCK_ACQUIRE();
+		if (!errcode) {
+			errcode = hci_internal_msg_get(p_hci_buffer, &msg_type);
+			MULTITHREADING_LOCK_RELEASE();
+		}
+		if (errcode) {
+			msg_type = 0;
+			return false;
+		}
 	}
 
+	bool retry_later = false;
 	if (msg_type == SDC_HCI_MSG_TYPE_EVT) {
-		event_packet_process(p_hci_buffer);
+		retry_later = event_packet_process(p_hci_buffer);
 	} else if (msg_type == SDC_HCI_MSG_TYPE_DATA) {
-		data_packet_process(p_hci_buffer);
+		retry_later = data_packet_process(p_hci_buffer);
 	} else {
 		__ASSERT(false, "sdc_hci_msg_type_t has changed. This if-else needs a new branch");
```

Comment:
could we get rid of this assert if we use a switch statement instead and make use of the right compiler warnings?